src/support/bits.h

```diff
@@ -66,14 +66,14 @@ int CountLeadingZeroes(T v) {
   return CountLeadingZeroes(typename std::make_unsigned<T>::type(v));
 }
 
-template <typename T>
-inline static T RotateLeft(T val, T count) {
+template <typename T, typename U>
```

Comment:
Derek, you didn't miss anything. The PR's came in at different times because I'm not yet spun up with git commands. I need some more experience.
You raised some questions which I probably should have considered more deeply. Shift and rotate semantics can differ between languages, implementations and instruction sets.
Regarding shift and rotate operators, I don't see any relationship between the types of the left and right operands. It seems wrong to me that they would need to be the same. I certainly don't see the need when looking at the WebAssembly design document or our C++ implementation.
I've now read the WebAssembly ast-semantics design document. It says the right hand operand (RHO) must be treated as an unsigned value and only the lower 5/6 bits are actionable. I don't see that a positive int (other.i32) will be semantically different than an unsigned. What about a negative other.i32. I'm not sure if it makes a semantic difference either but it does give me a headache thinking it through. Even if a negative shift count can be handled, I'm not sure if it's worth giving even one comparision/anding/negation to deal with that strange case.

 

    On Wednesday, May 11, 2016 9:20 PM, Derek Schuff <notifications@github.com> wrote:
 

 In src/support/bits.h:> @@ -66,14 +66,14 @@ int CountLeadingZeroes(T v) {
>    return CountLeadingZeroes(typename std::make_unsigned<T>::type(v));
>  }
>  
> -template <typename T>
> -inline static T RotateLeft(T val, T count) {
> +template <typename T, typename U>
Sorry I missed this before. I'm not sure having a second type is good here because they could end up being different sizes if we're not careful, and then the bit-twiddling doesn't work. If the warning is about negation of an unsigned value, maybe we could just cast it to signed for that expression instead.If we did that, I think we wouldn't need #474 either.—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub